riscv/mmu.cc

```diff
@@ -161,6 +161,65 @@ void mmu_t::load_slow_path(reg_t addr, reg_t len, uint8_t* bytes, uint32_t xlate
   }
 }
 
+void mmu_t::clean_inval(reg_t addr, bool clean, bool inval)
+{
+  reg_t paddr = addr;
+  bool page = false, guest = false, access = false;
+  bool has_gva;
+  reg_t gpa;
+
+  bool fault = true;
+  try {
+    paddr = translate(addr, 1, LOAD, 0);
+    fault = false;
+  } catch (trap_load_access_fault& t) {
+    access = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_page_fault& t) {
+    page = true;
+    has_gva = t.has_gva();
+  } catch (trap_load_guest_page_fault& t) {
+    guest = true;
+    gpa = t.get_tval2();
+  }
+
+  if (fault) {
+    try {
+      paddr = translate(addr, 1, FETCH, 0);
+    } catch (trap_instruction_access_fault& t) {
+      access = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_page_fault& t) {
+      page = true;
+      has_gva = t.has_gva();
+    } catch (trap_instruction_guest_page_fault& t) {
+      guest = true;
+      gpa = t.get_tval2();
+    }
+  }
+
+  if (access) {
```

Comment:
Two problems I see in this logic:

1. If LOAD detects an access fault, it tries again with FETCH. But `access` will still be true, so this will take a store access fault, when it should successfully perform the CBO.
2. This does not seem to check for STORE permissions at all? If the page is read-only, the operation should not be allowed (right?).

Reply:
Yeah. this is a logic error. I should take all of them to false if FETCH succeeds.